server/auth_callout.go

```diff
@@ -0,0 +1,329 @@
+		var err error
+		acc, err = s.LookupAccount(aname)
+		if err != nil {
+			s.Warnf("No valid account %q for auth callout request: %v", aname, err)
+			return false
+		}
+	} else {
+		acc = c.acc
+	}
+
+	// Check if we have been requested to encrypt.
+	var xkp nkeys.KeyPair
+	var xkey string
+	var pubAccXKey string
+	if !isOperatorMode && opts.AuthCallout != nil && opts.AuthCallout.XKey != _EMPTY_ {
+		pubAccXKey = opts.AuthCallout.XKey
+	} else if isOperatorMode {
+		pubAccXKey = acc.externalAuthXKey()
+	}
+	// If set grab server's xkey keypair and public key.
+	if pubAccXKey != _EMPTY_ {
+		// These are only set on creation, so lock not needed.
+		xkp, xkey = s.xkp, s.info.XKey
+	}
+
+	// Create a keypair for the user. We will expect this public user to be in the signed response.
+	// This prevents replay attacks.
+	ukp, _ := nkeys.CreateUser()
+	pub, _ := ukp.PublicKey()
+
+	reply := s.newRespInbox()
+	respCh := make(chan bool, 1)
+
+	processReply := func(_ *subscription, rc *client, racc *Account, subject, reply string, rmsg []byte) {
+		_, msg := rc.msgParts(rmsg)
+		// This signals not authorized.
+		// Since this is an account subscription will always have "\r\n".
+		if len(msg) == LEN_CR_LF {
+			respCh <- false
+			return
+		}
+		// Strip trailing CRLF.
+		msg = msg[:len(msg)-LEN_CR_LF]
+
+		// If we sent an encrypted request the response could be encrypted as well.
+		if xkp != nil && len(msg) > len(jwtPrefix) && !bytes.HasPrefix(msg, []byte(jwtPrefix)) {
+			var err error
+			msg, err = xkp.Open(msg, pubAccXKey)
+			if err != nil {
+				s.Warnf("Error decrypting auth callout response: %v", err)
+				respCh <- false
+				return
+			}
+		}
+
+		juc, err := jwt.DecodeUserClaims(string(msg))
+		if err != nil {
+			s.Warnf("Error decoding auth callout response: %v")
+			respCh <- false
+			return
+		}
+
+		// Make sure that the user is what we requested.
+		if juc.Subject != pub {
+			s.Warnf("Expected authorized user of %q but got %q", pub, juc.Subject)
+			respCh <- false
+			return
+		}
+
+		// Make sure correct issuer.
+		var issuer string
+		if opts.AuthCallout != nil {
+			issuer = opts.AuthCallout.Issuer
+		} else {
+			// Operator mode is who we send the request on unless switching accounts.
+			issuer = acc.Name
+		}
+		// By default issuer needs to match server config or the requesting account in operator mode.
+		if juc.Issuer != issuer {
+			if !isOperatorMode {
+				s.Warnf("Wrong issuer for auth callout response: %v")
+				respCh <- false
+				return
+			} else if !acc.isAllowedAcount(juc.Issuer) {
+				s.Warnf("Account %q not allowed for auth callout on %q", juc.Issuer, issuer)
+				respCh <- false
+				return
+			}
+		}
+		allowNow, validFor := validateTimes(juc)
+		if !allowNow {
+			c.Errorf("Outside connect times")
+			respCh <- false
+			return
+		}
+		allowedConnTypes, err := convertAllowedConnectionTypes(juc.AllowedConnectionTypes)
+		if err != nil && len(allowedConnTypes) == 0 {
+			respCh <- false
```

Comment:
Should something be done with the error?  Perhaps an account advisory?

Reply:
Changed to match behavior of this error in normal auth flow.

Not against advisories, but would want that in separate PR after some discussion and thought.